plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/grok/GrokProcessor.java

```diff
@@ -0,0 +1,131 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor.grok;
+
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.PathUtils;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.ingest.Data;
+import org.elasticsearch.ingest.processor.Processor;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.Map;
+
+public final class GrokProcessor implements Processor {
+
+    public static final String TYPE = "grok";
+    public static final String PATTERNS_CONFIG_DIRECTORY = "ingest/grok/patterns";
+
+    private final String matchField;
+    private final String matchPattern;
+    private final Grok grok;
+
+    public GrokProcessor(Grok grok, String matchField, String matchPattern) throws IOException {
+        this.matchField = matchField;
+        this.matchPattern = matchPattern;
+        this.grok = grok;
+        this.grok.compile(this.matchPattern);
+    }
+
+    @Override
+    public void execute(Data data) {
+        Object field = data.getProperty(matchField);
+        // TODO(talevy): handle invalid field types
+        if (field instanceof String) {
+            Map<String, Object> matches = grok.captures((String) field);
+            if (matches != null) {
+                matches.forEach((k, v) -> data.addField(k, v));
+            }
+        }
+    }
+
+    public static class Builder implements Processor.Builder {
+
+        private final Environment env;
+        private String patternsDirectory;
+        private String matchField;
+        private String matchPattern;
+        private Grok grok;
+
+        public Builder(Environment env) {
+            this.env = env;
+            this.grok = new Grok();
+        }
+
+        public void setMatchField(String matchField) {
+            this.matchField = matchField;
+        }
+
+        public void setMatchPattern(String matchPattern) {
+            this.matchPattern = matchPattern;
+        }
+
+        public void setPatternsDirectory(String patternsDirectory) {
```

Comment:
I don't think we should make the patterns dir configurable? Outside the ES_HOME directory ES has insufficient permissions to read files. I think the patterns dir should always be `$ES_HOME/config/ingest/grok/patterns`.